dot_parser.py

```diff
@@ -548,8 +548,13 @@ def parse_dot_data(s):
         tokens = graphparser.parseString(s)
         return list(tokens)
     except ParseException as err:
-        print(
-            err.line +
-            " "*(err.column-1) + "^" +
-            err)
+        if PY3 and (hasattr(ParseException, 'explain') and
```

Comment:
It is recommended to always check if PY2 (in case PY4 ever comes out :-)

Reply:
Hi @ankostis, thanks for reviewing my proposed changes, I really appreciate it. I will go by them one by one, as I find time.

About checking `PY3`. I am not sure if you are aware of the following:
- `PY3` is already used elsewhere in the code: 3 conditionals in `pydot.py` and 1 conditional in `dot_parser.py` use it already.
- Both files set it like this: `PY3 = sys.version_info >= (3, 0, 0)`, meaning Python 3 or higher. Maybe you were assuming that it was defined as in the [`six` library](https://github.com/benjaminp/six/blob/1.14.0/six.py#L37), as `sys.version_info[0] == 3`, meaning Python 3 _only_, in which case I would have agreed with your recommendation, but it is not.

You could still say:
- That the pydot definition of `PY3` is different than elsewhere and therefore confusing for pydot developers. In reply, I would say that I expect most developers will look up the `PY3` definition in the code when they run into it (as I did) and that any mistakes that result from any confusion may still get caught through code review or automated testing as well.
- That it might become more confusing once Python 4 is out, because people might then start to use `PY3` to mean 'not on Python 4 yet', which actually cannot be determined using our definition of `PY3`. In reply, I would say that the world may look very different by the time Python 4 comes out. The affected code lines may have been changed a dozen times between now and then already. Maybe they will have been deleted a long time before already, during a drop of support for Python 2. And again, I believe that the developers of that time will read the definition in the source code and change it as they see fit when adding conditionals for Python 3+4 support.

In total, I am not really against changing to `PY2`, but I also do not really see much benefit. In case you still recommend checking for `PY2`, can you at least agree that it should be done separate from this PR? Because it will require carefully changing 2 variable settings and flipping 4 conditionals (5 if you count this one) and that is all not really related to the subject of this PR.

Reply:
> That it might become more confusing once Python 4 is ...

Guido was clear on that, there won't ever be Python 4 (but otoh, Guido is not BDFL anymore:-) 

The whole point of checking PY2 (instead of PY3) is to sidestep our mental model of what it means a "regular python" - PY2 is the exception now.
In principle, you should be able to delete all PY2 conditionals, and the code should still work flawlessly in a modern OS.

In any case, i would always stay close to the official recommended procedures for porting PY2 code: https://docs.python.org/3/howto/pyporting.html#use-feature-detection-instead-of-version-detection

Reply:
> The whole point of checking PY2 (instead of PY3) is to sidestep our mental model of what it means a "regular python" - PY2 is the exception now.

Ok, I see, valid point. Still I hesitate to make that change in this PR now.

> In principle, you should be able to delete all PY2 conditionals, and the code should still work flawlessly in a modern OS.

Agreed, and this may be the easiest way out of this discussion. I am hoping there will be at least one more point release in the pydot 1.4.x branch that still supports Python 2, to provide those users with some of the important bug fixes that are currently awaiting approval. And then drop Python 2 and <3.5 support in a pydot 1.5.x version. Indeed that would clean up a lot of code and make life a lot easier. Parallel to 1.5.x, we could keep the 1.4.x branch alive for Python 2 users, backporting bug fixes as long as there are people willing to work on that.

> In any case, i would always stay close to the official recommended procedures for porting PY2 code: https://docs.python.org/3/howto/pyporting.html#use-feature-detection-instead-of-version-detection

Yes, I had read that section as well, but I still thought pydot's `PY3` definition was not so bad, at least not as bad as the first example.

Reply:
Maybe I should split this PR into the fix for the `TypeError` and the change/new 'feature' of using `ParseException.explain()`. That new feature would not fit in a point release anyway, so it would go to 1.5.x, so no need for `PY3` then anymore anyway. The `TypeError` fix is simple and can go in both branches 1.4.x and 1.5.x.

Reply:
I just created a separate PR #227 with just the minimal bug fix for `TypeError` for a bug fix/point release.